tx/transaction.go

```diff
@@ -292,22 +352,115 @@ func (t *Transaction) TestFeatures(supported Features) error {
 	return nil
 }
 
+// encodeTyped writes the canonical encoding of a typed transaction to w.
+func (t *Transaction) encodeTyped(w *bytes.Buffer) error {
+	w.WriteByte(t.Type())
+	return rlp.Encode(w, t.body)
+}
+
+// MarshalBinary returns the canonical encoding of the transaction.
+// For legacy transactions, it returns the RLP encoding. For typed
+// transactions, it returns the type the RLP encoding of the tx.
+func (tx *Transaction) MarshalBinary() ([]byte, error) {
+	if tx.Type() == LegacyTxType {
+		return rlp.EncodeToBytes(tx.body)
+	}
+	var buf bytes.Buffer
+	err := tx.encodeTyped(&buf)
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes the canonical encoding of transactions.
+// It supports legacy RLP transactions and typed transactions.
+func (t *Transaction) UnmarshalBinary(b []byte) error {
+	if len(b) > 0 && b[0] > 0x7f {
+		// It's a legacy transaction.
+		var data LegacyTransaction
+		err := rlp.DecodeBytes(b, &data)
+		if err != nil {
+			return err
+		}
+		t.setDecoded(&data, len(b))
+		return nil
+	}
+	// It's a typed transaction envelope.
+	inner, err := t.decodeTyped(b)
+	if err != nil {
+		return err
+	}
+	t.setDecoded(inner, len(b))
+	return nil
+}
+
 // EncodeRLP implements rlp.Encoder
 func (t *Transaction) EncodeRLP(w io.Writer) error {
-	return rlp.Encode(w, &t.body)
+	if t.Type() == LegacyTxType {
+		return rlp.Encode(w, &t.body)
+	}
+	buf := encodeBufferPool.Get().(*bytes.Buffer)
+	defer encodeBufferPool.Put(buf)
+	buf.Reset()
+
+	if err := t.encodeTyped(buf); err != nil {
+		return err
+	}
+	return rlp.Encode(w, buf.Bytes())
 }
 
 // DecodeRLP implements rlp.Decoder
 func (t *Transaction) DecodeRLP(s *rlp.Stream) error {
-	_, size, _ := s.Kind()
-	var body body
-	if err := s.Decode(&body); err != nil {
+	kind, size, err := s.Kind()
+
+	switch {
+	case err != nil:
 		return err
+	case kind == rlp.List:
+		// It's a legacy transaction.
+		var body LegacyTransaction
+		if err := s.Decode(&body); err != nil {
+			return err
+		}
+		*t = Transaction{body: &body}
+
+		t.cache.size.Store(thor.StorageSize(rlp.ListSize(size)))
+		return nil
+	case kind == rlp.String:
+		// It's a typed TX.
+		var b []byte
+		if b, err = s.Bytes(); err != nil {
+			return err
+		}
+		inner, err := t.decodeTyped(b)
+		if err == nil {
+			t.setDecoded(inner, len(b))
+		}
+		return err
```

Comment:
No strong feelings, but should this be changed ? 
```suggestion
		inner, err := t.decodeTyped(b)
		if err != nil { return err }
		t.setDecoded(inner, len(b))
		return nil
```